Updated shared sync collection refactoring

PiTrem opened the pull request:
first step of the sync-shared collection refactoring


- one table for collection 
- one table for  ACL 

TODO :

Commit:
added refact shared collection api and related frontend logic

Commit:
added refact shared collection api and related frontend logic

Commit:
- build shared collection tree in frontend
- modify sample serializer

Commit:
define id for root collection in ui

Commit:
add api for share whole collection

Commit:
- migrate collection data
- add function for collection tab

Commit:
added seed data to share collection between users

Commit:
add entity and collection actions for collection_acl

Commit:
add entity and collection actions for collection_acl

Commit:
add entity and collection actions for collection_acl

Commit:
remove sync collection scope and modify dependent methods

FabianMauz commented on app/api/helpers/collection_helpers.rb:
```diff
@@ -75,11 +72,11 @@ def fetch_collection_id_for_assign(prms = params, pl = 1)
-      Collection.joins(:sync_collections_users).where(
-        'sync_collections_users.id = ? and sync_collections_users.user_id in (?) and sync_collections_users.permission_level >= ?',
+    @collection = if current_collection['is_shared']
+      Collection.joins(:collection_acls).includes(:user).where(
```
would use ruby where method

FabianMauz reviewed (commented):
Part 1 - mostly looked at the ruby code, js part will come next

FabianMauz commented on app/packs/src/apps/mydb/collections/CollectionTree.js:
```diff
@@ -94,73 +99,37 @@ export default class CollectionTree extends React.Component {
-          My shared collections
+    return (
+      <div>
+        <div style={{ display: '' }}>
```
move style to css classes

FabianMauz commented on app/packs/src/apps/mydb/collections/CollectionTree.js:
```diff
@@ -94,73 +99,37 @@ export default class CollectionTree extends React.Component {
-    )
+  unsharedSubtrees() {
+    let roots = this.state.unsharedRoots;
+    roots = roots.filter(function (item) { return !item.isNew })
```
would stick to lambda approach as above: roots.filter(item => !item.isNew)

FabianMauz commented on app/packs/src/apps/mydb/collections/CollectionTree.js:
```diff
@@ -187,47 +156,74 @@ export default class CollectionTree extends React.Component {
+
+    let subTreeLabels = (
+      <div className="tree-view">
+        <div id="synchron-home-link" className="title" style={{backgroundColor:'white'}}
```
move style to css

FabianMauz commented on app/packs/src/apps/mydb/collections/CollectionTree.js:
```diff
@@ -187,47 +156,74 @@ export default class CollectionTree extends React.Component {
-  remoteSyncInSubtrees() {
-    let { syncInRoots, syncCollectionVisible } = this.state
-    syncInRoots = this.removeOrphanRoots(syncInRoots)
+  sharedByMeSubtrees() {
```
would add render to name to make clear something is rendered

FabianMauz commented on app/packs/src/apps/mydb/collections/CollectionTreeStructure.js:
```diff
@@ -0,0 +1,67 @@
+    let label = `by ${collection.shared_by.initials}`;
+    let user = {};
+    let uid = -1;
+    let sharedCollections = collection.collection_acls.filter(acl => (acl.user.id === currentUser.id ));
```
i think that are not collections but acls of collections.

FabianMauz commented on app/packs/src/apps/mydb/collections/CollectionTreeStructure.js:
```diff
@@ -0,0 +1,67 @@
+    if (sameSharedTo) {
+      children.forEach(c => sameSharedTo.children.push(c))
+    } else {
+      let sharedCollection = {}
```
here a new collection is created. I would extract that functionality to reduce complexity
and make it usable for the next function with also use this logic

FabianMauz commented on app/packs/src/apps/mydb/collections/CollectionTreeStructure.js:
```diff
@@ -0,0 +1,67 @@
+    let label = ''
+    let user = {}
+    let uid = -1;
+    collection.collection_acls.forEach((acl) => {
```
extract this block as function to make it usable for above function

FabianMauz commented on app/packs/src/fetchers/CollectionsFetcher.js:
```diff
@@ -49,6 +49,96 @@ export default class CollectionsFetcher {
   }
 
+  static fetchMyRoots() {
+    let promise = fetch('/api/v1/temp_collections/all', {
```
just return the fetch statement, no need to have line 64

FabianMauz commented on app/packs/src/fetchers/CollectionsFetcher.js:
```diff
@@ -49,6 +49,96 @@ export default class CollectionsFetcher {
     return promise;
   }
```
fetchMyRoots( ) , fetchSharedWithMeRoots ( ), fetchMyRoots ( ), fetchSharedWithMeRoots( ), fetchMyRoots ( ) and fetchSharedWithMeRoots ( ) all have the same structure/logic. Only the adress is different. I would extract the structure and have general api caller method with the address as a parameter

FabianMauz reviewed (commented):
Part 2 - looked at the js code

Commit:
add rspec for shared collcetion api

Commit:
refactor move and assign to collection api

Commit:
update elementStore and fetcher for move or assign collection

endevor-music commented:
A couple of (I think) significant front end issues I noticed between current main and feature branch:

Can someone confirm reproducibility of the following, please? @mehreenmansur @PiTrem 

1) Change in Collection UI

feature branch:
- The All collection does not appear, despite the URL pointing to `mydb/collection/all`, and the chemotion-repository.net collection is not visible either. Both appear in the collections list, however.
![Screenshot 2023-04-19 142749](https://user-images.githubusercontent.com/67055123/233075589-9e33f693-cfd4-41da-84e2-163e8d53cc40.png)

2) Change in URL

- When clicking on a collection, the URL changes from `collection` to `temp_collections`.
![Screenshot 2023-04-19 142810](https://user-images.githubusercontent.com/67055123/233075591-5b64ae88-1d87-447f-8ae9-61281b094a1e.png)

Furthermore, when changing `temp_collections` back to `collection` in the URL, the following `TypeError` occurs.
![Screenshot 2023-04-19 145817](https://user-images.githubusercontent.com/67055123/233082473-d511e546-8e9e-40ef-a526-ed2c7fe60ea7.png)

- Is the change in URL intentional (I would hope not!), or a quirk of the new API?
- If the change in URL is intentional, having `temp_collections` as a name invokes a sense of, well, the collection being temporary. I think a rename, to something like `shared_collections`, could be required.

main, for comparison:
![Screenshot 2023-04-19 143012](https://user-images.githubusercontent.com/67055123/233075479-ff498c0d-53c3-4e50-8770-7ac43c056427.png)

- I shall request a change for this in the code - but the shared with me collection header needs to be updated. 
![Screenshot 2023-04-19 150828](https://user-images.githubusercontent.com/67055123/233084563-82d3c2d3-65d4-47bb-bd2d-866fa798110a.png)

FInally - an update button exists for the user (User A) to update collections they have shared with others (Users B, C). But no such update button exists for updating the list of collections that others (Users B,C) have shared with a user (User A) - is it possible to implement this as well? Having similar functionality for similar features is ideal from a user point of view.

endevor-music commented on app/packs/src/apps/mydb/collections/SharedWithMeCollections.js:
line 61 needs to be changed

mehreenmansur commented:
> temp_collections

@endevor-music The temp_collection link is intentional for now we tend to keep it. Once the feature testing is done we will replace it.

Commit:
update temp collection endpoints

Commit:
update temp collection endpoints

Commit:
rename share collection button

Commit:
- add unique index
- move, assign api update

Commit:
refactor synced collection methods

Commit:
add collection migration data

PiTrem committed:
rm shared collection navigation

- rm scollectionShow

- ref Aviator navigation to collection/element to routesUtils

PiTrem committed:
pass collection  object to subtree

PiTrem committed:
replace more Aviator.navigate by AviatorNavigation

PiTrem committed:
deprecate isShared

Commit:
updated:
- collection tree order
- research plan details show on share collection

PiTrem committed:
WIP deprecate is_shared

PiTrem committed:
Collection permission level: define and describe the levels

PiTrem committed:
remove is_shared params from API endpoints

DRY assignment of element to collection + 'All' collection

TODO ? move assignment to collection + 'All' logic to use-case or concern ?

TODO clean test

PiTrem committed:
remove is_shared params from API endpoints

DRY assignment of element to collection + 'All' collection

TODO ? move assignment to collection + 'All' logic to use-case or concern ?

TODO clean test

PiTrem committed:
fix Constant name

Commit:
- update ownership transfer logic
- remove sync collection endpoints

PiTrem committed:
clean spec

PiTrem committed:
clean db/schema

PiTrem committed:
rename endpoint to get all collections

PiTrem committed:
rm unused collection fetcher/action

rm forgotten isShared params

rename collection fetcher to get all own collections

PiTrem committed:
fix api get elements for collection

PiTrem committed:
UI navigation to a collection All or by Id

avoid unecessary fetching to server

PiTrem committed:
build collection tree in store

PiTrem committed:
define collection scope owned_by(user_ids) and shared_with(userids)

PiTrem committed:
Collection and CollectionAcl entities

PiTrem committed:
use collectiontree

PiTrem committed:
dedicated collection entity

PiTrem committed:
dprecate fetchSharedCollection

PiTrem committed:
get collection by id: authorize user

PiTrem committed:
use same endpoint to delte a share collection

collection owner and shared-with user can both delete the access to the
collection

PiTrem committed:
remove redundant share_collection api endpoint

that are done by collection api

fix authorization for Delete ACL
add authorization for put ACL

Commit:
- fix labels for collections shared
- remove my shared collection management

PiTrem committed:
fix navigation

PiTrem committed:
add level details to entities

PiTrem committed:
set UIstore current collection from collection store

PiTrem committed:
collection-store : rename descendants children

PiTrem committed:
JS Collection Model

PiTrem committed:
mv collection properties to collection model

PiTrem committed:
rm duplicate user entity

PiTrem committed:
CollectionEntity: return acl only for current user

PiTrem committed:
ContextAction Btn: move checks on collection to collection model

PiTrem committed:
Collection.rb fix owned_by

PiTrem committed:
Manageing Action: refactor collection-tree option

TODO filter: according to collection permission

PiTrem committed:
collection entity: serialize position

PiTrem committed:
move more logic to collection model
test: collection.spec.js, collectionFactory (vanilla)

PiTrem committed:
fix permission managing action

PiTrem committed:
rename collection method

 alias permission level

PiTrem committed:
refactor collection tree in Export-Collection Modal

PiTrem committed:
move collection list option to collection store

PiTrem committed:
Collection  navigation: fix initial loading

ptrxyz committed:
Update collection_acl_entity.rb

Fixed a typo, UserEntity -> UserSimpleEntity

mehreenmansur committed:
refactor transfer of ownership (#1332)

* update handle transfer of ownership

* update take ownership usecase method

* - add pass ownership method
- update api to pass collection id

---------

Co-authored-by: Mehreen <mehreen.mansur>

mehreenmansur committed:
Element tag refactor (#1342)

* migration of  element tag data
 - fetch actual collection ids
 - destroy if no element found


* update element tag concern


* date sql for report helpers


* element_tag to include all collections


---------

Co-authored-by: Mehreen <mehreen.mansur>
Co-authored-by: PiTrem <pierre.tremouilhac@kit.edu>

Commit:
fix read-only permission issues

PiTrem committed:
fix(rebase): collection entity

PiTrem committed:
fix(rebase): showDetails fn

Commit:
fixed total elements issue on navigation

Commit:
update send messages method for comment function

PiTrem committed:
chores(API): clean sync_collection remnant

Commit:
add share_collection_api
show number of sub-folders [1579]

Commit:
check permission level
update report_api

Commit:
remove from collection issue [1033]

Commit:
update error message format

mehreenmansur committed:
update element tag data migration

mehreenmansur committed:
associated samples delete from all collection
fix label generation in modal

mehreenmansur committed:
delete elements from all collections on transfer of ownership

mehreenmansur committed:
transfer of ownership: delete frm other collections of previous owner

mehreenmansur committed:
fix for sync permission level issue [1755]

mehreenmansur committed:
updated permission levels params pass for acl collections

mehreenmansur committed:
update helper join element class methods

mehreenmansur committed:
add collection_acl migration with cell line detail level

mehreenmansur committed:
update for remove from collection issue [1033]

mehreenmansur committed:
update method of assign or move to collection

mehreenmansur committed:
remove sync_collection_user dependencies

PiTrem committed:
fixup: mv ACl table creation migration back in time

PiTrem committed:
fixup: cleanup db/schema / fix migration timetravel

ashwiniHerle committed:
adding schema

Commit:
fix: the issue with the acceptance of ownership
refactor: code

Commit:
style: improve the TakeOwnership button

ashwiniHerle committed:
Revert "adding schema"

This reverts commit 2552644dda258ba381cac12b5b7043c4ff4f1b34.

ashwiniHerle committed:
fix: collection tree not to collapse on selecting another collection (#1988)

Commit:
fix: issue of ElementTag not being displayed
fix: count of My Collections in the ElementTag
fix: collection names being disabled for own collections
fix: issue with creation of ElementTag data
fix: issue with update of ElementTag data

Commit:
fix: the text under "Collections shared with me" tab in collection management

ashwiniHerle committed:
fix: broken cellline count (#1992)

Commit:
fix: accepting ownership by new owner removes the original owner from the collection rights. Now it will only switch the sharer and shared roles when transfer of ownership is accepted

Commit:
fix: shared collection created when no label entered
refactor: rubocop warning

Commit:
fix: delete shared collection not working

Commit:
feat: automatically lower the permission level as (pass ownership - 1) i.e. set the permission level to 4 for the new owner

Commit:
fix: sharing collections with groups not working (#188)

Commit:
fix: collections shared with group must not be deletable by the group member
- blocked in FE
- blocked in BE
- only a user is able to delete the share collection access if the collection has been shared with the user

Commit:
fix: collection owner not being able to update elements

Commit:
fix: sharing elements not working when the collection has been shared with that user
